#Take Home Exam
Name: Jeny Johnmary (i6197039) |
Date: 09/02/2022

## Introduction

Research question:
<brb>

***How have the Government Response Stringency Index influenced the new confirmed cases of COVID-19 per million in different Western European countries in 2020?***
<brb>

* Stringency Index:
 * value from 0 - 100, composite measure based on nine response indicators (school closure, workplace closure, travel bans, etc.

<brb>

* New confirmed cases
 * per million 




#Data analysis:

In [ ]:
#import the essential packages I need for my analysis
import altair as alt
import pandas as pd
import ipywidgets as widgets
from ipywidgets import interact
from ipywidgets import fixed

In [ ]:
#read the COVID data and stored in variable
Covid = pd.read_csv("https://covid.ourworldindata.org/data/owid-covid-data.csv", parse_dates=['date'])
Covid

iso_code  ... excess_mortality_cumulative_per_million
0           AFG  ...                                     NaN
1           AFG  ...                                     NaN
2           AFG  ...                                     NaN
3           AFG  ...                                     NaN
4           AFG  ...                                     NaN
...         ...  ...                                     ...
160232      ZWE  ...                                     NaN
160233      ZWE  ...                                     NaN
160234      ZWE  ...                                     NaN
160235      ZWE  ...                                     NaN
160236      ZWE  ...                                     NaN

[160237 rows x 67 columns]

In [ ]:
#checked the unique values of de continent columns to see if there are strange values
Covid.continent.unique()

array(['Asia', nan, 'Europe', 'Africa', 'North America', 'South America',
       'Oceania'], dtype=object)

In [ ]:
#only included Europe and the relevant columns in a new dataframe
included_continent = ['Europe']
included_columns = ['iso_code', 'continent', 'location', 'date', 'stringency_index', 
                    'new_cases_per_million', 'population', 'median_age']
Covid_EU = Covid.loc[Covid.continent.isin(included_continent) & (Covid.date.dt.year == 2020), included_columns]
Covid_EU

iso_code continent  ... population median_age
1441        ALB    Europe  ...  2872934.0       38.0
1442        ALB    Europe  ...  2872934.0       38.0
1443        ALB    Europe  ...  2872934.0       38.0
1444        ALB    Europe  ...  2872934.0       38.0
1445        ALB    Europe  ...  2872934.0       38.0
...         ...       ...  ...        ...        ...
155110      VAT    Europe  ...      812.0        NaN
155111      VAT    Europe  ...      812.0        NaN
155112      VAT    Europe  ...      812.0        NaN
155113      VAT    Europe  ...      812.0        NaN
155114      VAT    Europe  ...      812.0        NaN

[15315 rows x 8 columns]

In [ ]:
#checked the unique values of locations in EU, to find if there are strange values
Covid_EU.location.unique()

array(['Albania', 'Andorra', 'Austria', 'Belarus', 'Belgium',
       'Bosnia and Herzegovina', 'Bulgaria', 'Croatia', 'Cyprus',
       'Czechia', 'Denmark', 'Estonia', 'Faeroe Islands', 'Finland',
       'France', 'Germany', 'Gibraltar', 'Greece', 'Hungary', 'Iceland',
       'Ireland', 'Isle of Man', 'Italy', 'Kosovo', 'Latvia',
       'Liechtenstein', 'Lithuania', 'Luxembourg', 'Malta', 'Moldova',
       'Monaco', 'Montenegro', 'Netherlands', 'North Macedonia', 'Norway',
       'Poland', 'Portugal', 'Romania', 'Russia', 'San Marino', 'Serbia',
       'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland',
       'Ukraine', 'United Kingdom', 'Vatican'], dtype=object)

In [ ]:
#made a new dataframe with only the counrties of West EU
included_countries = ['Ireland', 'United Kingdom', 'Germany', 'Netherlands', 'Belgium',
                      'Luxembourg', 'France', 'Austria', 'Slovenia']
included_columns = ['iso_code', 'continent', 'location', 'date', 'stringency_index', 
                    'new_cases_per_million', 'population', 'median_age']
Covid_West_EU = Covid_EU.loc[Covid_EU.location.isin(included_countries) & (Covid_EU.date.dt.year == 2020), included_columns]
Covid_West_EU

iso_code continent  ...  population median_age
9313        AUT    Europe  ...   9043072.0       44.4
9314        AUT    Europe  ...   9043072.0       44.4
9315        AUT    Europe  ...   9043072.0       44.4
9316        AUT    Europe  ...   9043072.0       44.4
9317        AUT    Europe  ...   9043072.0       44.4
...         ...       ...  ...         ...        ...
151063      GBR    Europe  ...  68207114.0       40.8
151064      GBR    Europe  ...  68207114.0       40.8
151065      GBR    Europe  ...  68207114.0       40.8
151066      GBR    Europe  ...  68207114.0       40.8
151067      GBR    Europe  ...  68207114.0       40.8

[2924 rows x 8 columns]

In [ ]:
#checked the missing values in the datafram that I am interested in
Covid_West_EU.isnull().sum()

iso_code                  0
continent                 0
location                  0
date                      0
stringency_index          0
new_cases_per_million    40
population                0
median_age                0
dtype: int64

In [ ]:
#only 39 rows of 2924 had missing values (only new cases), so I dropped them 
Covid_West_EU_cleaned = Covid_West_EU.dropna()
Covid_West_EU_cleaned

iso_code continent  ...  population median_age
9313        AUT    Europe  ...   9043072.0       44.4
9314        AUT    Europe  ...   9043072.0       44.4
9315        AUT    Europe  ...   9043072.0       44.4
9316        AUT    Europe  ...   9043072.0       44.4
9317        AUT    Europe  ...   9043072.0       44.4
...         ...       ...  ...         ...        ...
151063      GBR    Europe  ...  68207114.0       40.8
151064      GBR    Europe  ...  68207114.0       40.8
151065      GBR    Europe  ...  68207114.0       40.8
151066      GBR    Europe  ...  68207114.0       40.8
151067      GBR    Europe  ...  68207114.0       40.8

[2884 rows x 8 columns]

In [ ]:
chart_cases_time= alt.Chart(Covid_West_EU_cleaned).mark_line().encode(
    x='date',
    y='new_cases_per_million',
    color ='location'
).interactive()
chart_cases_time

alt.Chart(...)

In [ ]:
#made some descriptives of the interested variables per country
Descr_POP = Covid_West_EU_cleaned.groupby('location')['population'].describe()
Descr_MedAge = Covid_West_EU_cleaned.groupby('location')['median_age'].mean()
Descr_SI = Covid_West_EU_cleaned.groupby('location')['stringency_index'].describe()
Descr_NCPM = Covid_West_EU_cleaned.groupby('location')['new_cases_per_million'].describe()

In [ ]:
Descr_POP

count        mean  std  ...         50%         75%         max
location                                ...                                    
Austria         311.0   9043072.0  0.0  ...   9043072.0   9043072.0   9043072.0
Belgium         332.0  11632334.0  0.0  ...  11632334.0  11632334.0  11632334.0
France          340.0  67422000.0  0.0  ...  67422000.0  67422000.0  67422000.0
Germany         340.0  83900471.0  0.0  ...  83900471.0  83900471.0  83900471.0
Ireland         307.0   4982904.0  0.0  ...   4982904.0   4982904.0   4982904.0
Luxembourg      307.0    634814.0  0.0  ...    634814.0    634814.0    634814.0
Netherlands     309.0  17173094.0  0.0  ...  17173094.0  17173094.0  17173094.0
Slovenia        302.0   2078723.0  0.0  ...   2078723.0   2078723.0   2078723.0
United Kingdom  336.0  68207114.0  0.0  ...  68207114.0  68207114.0  68207114.0

[9 rows x 8 columns]

In [ ]:
Descr_MedAge

location
Austria           44.4
Belgium           41.8
France            42.0
Germany           46.6
Ireland           38.7
Luxembourg        39.7
Netherlands       43.2
Slovenia          44.5
United Kingdom    40.8
Name: median_age, dtype: float64

In [ ]:
Descr_SI

count       mean        std    min    25%    50%    75%    max
location                                                                      
Austria         311.0  55.352926  19.659155  11.11  37.96  50.00  75.00  82.41
Belgium         332.0  56.277831  19.369458  11.11  50.00  58.33  65.74  81.48
France          340.0  57.679706  23.590948   5.56  46.30  51.85  76.85  87.96
Germany         340.0  55.725882  19.177287   5.56  49.54  59.72  64.35  82.41
Ireland         307.0  66.270782  20.330727  11.11  52.31  72.22  81.48  90.74
Luxembourg      307.0  50.941010  16.751157   0.00  41.67  47.22  60.19  79.63
Netherlands     309.0  58.314693  16.714436   5.56  50.93  62.04  71.30  78.70
Slovenia        302.0  60.581358  19.732109  11.11  47.22  50.00  81.48  89.81
United Kingdom  336.0  61.807857  21.340004   8.33  64.35  67.59  73.15  79.63

In [ ]:
Descr_NCPM

count        mean         std  ...      50%        75%       max
location                                       ...                              
Austria         311.0  128.294537  209.588011  ...  27.7560  141.32400  1060.038
Belgium         332.0  167.402102  313.603027  ...  50.2050  161.31775  2056.423
France          340.0  118.991729  203.993614  ...  32.5860  167.63800  1553.009
Germany         340.0   61.239541   99.240825  ...  15.6910   57.43100   430.903
Ireland         307.0   59.995990   67.460107  ...  42.9470   80.57550   344.779
Luxembourg      307.0  238.162560  499.997307  ...  66.1610  196.12050  3098.545
Netherlands     309.0  150.464362  198.804716  ...  38.1410  272.86900   759.153
Slovenia        302.0  194.579457  329.657942  ...  16.1155  208.06050  1256.541
United Kingdom  336.0  108.730390  151.452140  ...  37.5110  188.71900   819.944

[9 rows x 8 columns]

In [ ]:
#now I wanted to visualize the data in an interactive way

#firstly, I defined how to select the rows of locations from the df

def select_country(location, df):
  return df.loc[df['location'] == location, ]

#then the definition for the vizualization
def vizualize_country(location, df):

  filtered_data = select_country(location, df)

#then I created the two graphs for stringency_index and new_cases_per_million
  chart_stringency = alt.Chart(filtered_data).mark_point().encode(
    x='date',
    y='stringency_index',
    color = 'location'
  )
  
  chart_new_cases = alt.Chart(filtered_data).mark_point().encode(
    x='date',
    y='new_cases_per_million',
    color = 'location'
  ) 

#the two graphs will be displayed
  display(chart_stringency) 
  
  display(chart_new_cases)

In [ ]:
#creating a list for the dropdown menu I will use in the following cell
country_list = Covid_West_EU_cleaned['location'].unique()

In [52]:
#to present the interactive graph I created a nice border
output_interactive_graph = widgets.Output(layout={'border': '2px solid blue'})

##within the output border I presented a text and the two graphs
###with the interactive dropdown menu of the locations in West EU
with output_interactive_graph:
  print("Choose a country in West EU to see the effect of Government Response") 
  print("Stringency Index on confirmed new COVID-19 cases per million in 2020")
  interact(vizualize_country, location=country_list, df=fixed(Covid_West_EU_cleaned))

output_interactive_graph 

Output(layout=Layout(border='2px solid blue'))

# Conclusion
* Population differed a lot between the countries in West-EU.
  * Median age is in every country about 40 y/o.
* Luxembourg had the lowest stringency index (50.94) and the highest new confirmed COVID-19 cases per million (238.16).
* Ireland had the highest stringency index (66.27) and the lowest new confirmed COVID-19 cases per million (59.99).
* Rest of the countries had a quite similar stringency index (about 55), but the new confirmed COVID-19 cases differed a lot from each other.

<brb>

An overall conclusion (at this moment) will be that a signigicant higher stringency index will lead to an significant lower new confirmed COVID-19 cases per million. However there are exceptions in some countries. 
